src/Microsoft.DocAsCode.Build.Engine/DocumentBuildContext.cs

```diff
@@ -193,6 +201,111 @@ private List<string> ResolveByExternalReferencePackages(List<string> uidList, Co
             return list;
         }
 
+        private async Task<List<string>> ResolveByXRefServiceAsync(List<string> uidList, ConcurrentDictionary<string, XRefSpec> externalXRefSpec)
+        {
+            if (XRefServiceUrls == null || XRefServiceUrls.Length == 0)
+            {
+                return uidList;
+            }
+            int needQueryCount = uidList.Count;
+            var unresolvedUidList = new List<string>();
+            int pieceSize = 1000;
+            foreach (string requestUrl in XRefServiceUrls)
+            {    
+                for(int i = 0; i < uidList.Count; i+= pieceSize)
+                {
+                    List<string> smallPiece = uidList.GetRange(i, Math.Min(pieceSize, uidList.Count - i));
+                    var result = await QueryByHttpRequestAsync(requestUrl, smallPiece);
+                    bool isSuccess = result.Item1;
+                    List<XRefSpec> xsList = result.Item2;
+                    if (isSuccess)
+                    {
+                        for (int j = 0; j < xsList.Count; j++)
+                        {
+                            if (xsList[j] != null)
+                            {
+                                externalXRefSpec.AddOrUpdate(smallPiece[j], xsList[j], (_, old) => old + xsList[j]);
+                            }
+                            else
+                            {
+                                unresolvedUidList.Add(smallPiece[j]);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        unresolvedUidList.AddRange(smallPiece);
+                    }
+                }
+                Logger.LogInfo($"{uidList.Count - unresolvedUidList.Count} uids found in {requestUrl}.");
+                uidList = new List<string>(unresolvedUidList);
```

Comment:
less garbage

---
In reply to: [127868218](https://github.com/dotnet/docfx/pull/1837#discussion_r127868218) [](ancestors = 127868218,127660104)